### PR TITLE
Add pass to fuse pack operations into for loop

### PR DESCRIPTION
**Pull request by yzhang93**

### CODE REVIEW

**Comment by Abhishek-Varma** [comment]
This looks okay to me.
I was trying to think about how to have lesser footprint - since we currently have two cases `(fill, forall)` and `(pack, for)` - I was thinking if we can instead add an option within the previous similar pass of this nature to make the logic based on a bool `useSCFFor` and accordingly we rename the pass as `--iree-amdaie-specific-fuse` (I'm clearly not good with names. :P)

Not adding a blocker on this though - just being vocal about another approach that (might?) ensure seemingly less footprint.

**Comment by yzhang93** [comment]
> This looks okay to me. I was trying to think about how to have lesser footprint - since we currently have two cases `(fill, forall)` and `(pack, for)` - I was thinking if we can instead add an option within the previous similar pass of this nature to make the logic based on a bool `useSCFFor` and accordingly we rename the pass as `--iree-amdaie-specific-fuse` (I'm clearly not good with names. :P)
> 
> Not adding a blocker on this though - just being vocal about another approach that (might?) ensure seemingly less footprint.

Thanks for your suggestion and I think it's reasonable. Currently I would keep these as two separate passes, because the implementation details are different and we want particular combinations of the producer and the loops. Even though we can add them to the same pass, we'll need to have separate implementations.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:40]: This is looking for a fill op not a pack op?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:90]: This is misleading, the producer of slice is not necessarily a pack op, we are collecting all the sliceops in the loop and they can come from any kind of producer, and we are trying to fuse all of them here, not just pack ops

**Review by nirvedhmeshram — CHANGES_REQUESTED** [review]
Please change the description to clarify that this pass fuses all the producers in the for loop and not just the pack ops.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:40]: Thanks for catching this bug. This should be a tensor::PackOp.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:51]: THis is kind of backwards. If you want to fuse the producers of the pack inside the loop, I am assuming you are only looking at lhs and rhs operands of the compute operation. 

What you should do instead is look for the compute op, if its operands are slices and the source of the slices are pack, fuse them in... 

Though, i have been thinking about this. A cleaner way to do this would be to just do the packing within the scf.for op and then just hoist the pack for the output operands out... conversely you should "hoist" the unpack operation out as well (but it goes after the loop). I havent fully worked this out though. I need some time to play with the transform dialect script you shared :P

**Review by MaheshRavishankar — COMMENTED** [review]
I have some comments, but this pass is inherently strange. This is a means to and end and is not a long term thing... Lets evaluate benefit before landing it.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:37]: Thanks for taking a look into this. I'll push up another PR that using this pass to get to what AIR and AIE hardware requires. We can then evaluate the benefit of this pass. At the same time, I modified the logic a little bit to match what you described.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:84]: Thanks for pointing this out. I modified the logic a little bit, but still constrains the producer as a tensor::PackOp to avoid undesired fusion...

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:57]: You can just do 
```
for () {
  auto sliceOp = dyn_cast_or_null<tensor::ExtractSliceOp>(operand.getDefiningOp());
  if (sliceOp && isa_and_non_null<tensor::PackOp>(sliceOp.getSource())) {
     sliceOps.push_back(sliceOp);
  }
}
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp:76]: You can drop this check completely with the above check.

**Review by MaheshRavishankar — APPROVED** [review]
Ok, left a comment which should make this easier. After that this looks good. APproving to unblock landing this.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp]: Thanks. Done.

**Review by nirvedhmeshram — APPROVED** [review]

**Pull request merged**
